### PR TITLE
[ENH] Zero-Inflated compositor, Poisson, and Negative Binomial distributions

### LINKED ISSUE #554: [ENH] Zero-inflated distributions
Zero-inflated distributions are important in intermittent demand forecasting, so it would be nice to have these as explicit distributions, and/or a compositor.

Based on discussion with @tingiskhan in https://github.com/sktime/sktime/pull/8438.

I can think of three high-priority ones, for discussion:

- [ ] zero-inflated Poisson
- [ ] zero-inflated Negative Binomial
- [ ] zero-inflated compositor, i.e., inflates any distribution with zeros, e.g., `ZeroInflated(any_distribution, zero_rate=0)`. This should have analytic formulae for all of the distribution defining methods.

A possible approach - though challenging - might be to write the zero-inflated compositor *first*, and then define zero-inflated Poisson and NB by delegation, using `_DelegatedDistribution`.

### PR DESCRIPTION
**Pull request by marrov**
> **Note:** This PR description was created with assistance from Claude Sonnet 4.5. The implementation was developed with Claude Opus 4.5, followed by a manual review and automated code review by GPT 5.2 Codex.

#### Reference Issues/PRs
Closes #554 

#### What does this implement/fix? Explain your changes.

I'm currently working with `XBGoostLSS` and need to be able to use the zero-inflated distributions that this library already supports (see [here](https://statmixedml.github.io/XGBoostLSS/distributions/)). To this end, this PR adds support for **zero-inflated distributions** to skpro, which are essential for modeling count data with excess zeros (common in retail, healthcare, and insurance).

**Three new distribution classes:**

1. **`ZIPoisson`** - Zero-Inflated Poisson distribution
   - Standalone implementation extending `_ScipyAdapter`
   - Optimized for zero-inflated Poisson modeling

2. **`ZINB`** - Zero-Inflated Negative Binomial distribution
   - Standalone implementation extending `_ScipyAdapter`
   - Handles overdispersed count data with excess zeros
   - Includes numerical stability safeguards (epsilon clipping for `probs` conversion)

3. **`ZeroInflated`** - Generic zero-inflated compositor
   - Wraps any `BaseDistribution` to create zero-inflated versions
   - Enables zero-inflation for arbitrary distributions (e.g., `ZeroInflated(pi=0.3, distribution=Gamma(...))`)
   - Dynamically updates capabilities based on base distribution
   - Includes support/validation warnings for non-negative base distributions

**Key Features:**

- **Complete probability functions:** PMF/PDF, log-PMF/log-PDF, CDF, PPF, mean, variance
- **Comprehensive parameter validation:** Enforces `pi ∈ [0, 1)`, `mu > 0`, `alpha > 0` with clear error messages
- **Broadcasting support:** Works with scalar and 2D array parameters with proper shape validation
- **XGBoostLSS integration:** Both ZINB and ZIPoisson are fully supported in the `XGBoostLSS` wrapper with parameter mapping from xgboostlss to skpro parametrizations
- **Array indexing:** Implements `_iloc` and `_iat` for the compositor to support slicing operations

**Design Rationale:**

The implementation follows skpro's established pattern of providing both **specialized standalone implementations** (for performance and common use cases) and **generic compositors** (for flexibility):

- **Standalone** (`ZINB`, `ZIPoisson`): Direct implementations optimized for common statistical models with cleaner API and xgboostlss integration
- **Compositor** (`ZeroInflated`): Enables zero-inflation for any distribution, useful for custom models and research applications

This mirrors patterns like `TruncatedNormal` vs. `TruncatedDistribution`, and `LogNormal` vs. `TransformedDistribution`.

**Testing:**

- 61 tests covering all three distributions
- Standalone distribution tests: negative values, mean/variance properties, CDF behavior
- Compositor equivalence tests: `ZeroInflated(Poisson)` matches `ZIPoisson`
- Parameter validation tests: 14 tests for invalid parameters (negative, out-of-range, shape mismatches)
- Boundary condition tests: `pi=0` (reduces to base distribution), `pi→1` (mostly zeros)
- XGBoostLSS integration test script with synthetic data fitting (local only, this was not committed)

All tests are check to be passing locally.

#### Does your contribution introduce a new dependency? If yes, which one?

No new core dependencies.

#### What should a reviewer concentrate their feedback on?

- [ ] **Mathematical correctness** of zero-inflated probability formulas (PMF, CDF, PPF, moments)
- [ ] **Parameter validation** completeness and error messages
- [ ] **Numerical stability** of ZINB parameter conversion in `xgboostlss.py` (epsilon clipping approach)
- [ ] **Compositor design**: Does the `ZeroInflated` class properly handle negative-support distributions? (Currently clamps to 0 for x<0 and documents this limitation)
- [ ] **API consistency** with other skpro distributions
- [ ] **Test coverage**: Are there edge cases we should add?
- [ ] **Documentation clarity** - especially the distinction between standalone vs. compositor usage

#### Did you add any tests for the change?

**Yes,** comprehensive test suite added in `skpro/distributions/tests/test_zero_inflated.py`:

- **Basic property tests** (18 tests): negative values, mean≥0, variance≥0, CDF at zero for ZIPoisson, ZINB, ZeroInflated
- **Compositor equivalence tests** (2 tests): Verify `ZeroInflated(Poisson)` ≡ `ZIPoisson` and `ZeroInflated(NegativeBinomial)` ≡ `ZINB`
- **Parameter validation tests** (14 tests): Invalid pi (negative, ≥1), invalid mu (≤0), invalid alpha (≤0), invalid shapes
- **Boundary condition tests** (6 tests): pi=0 reduces to base, pi→1 gives high zero probability
- *Integration test script** (`scripts/test_xgboostlss_zinb.py`): Imports, creation, sampling, xgboostlss fitting with synthetic data

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
* there is a pull request for the compositor already, here: https://github.com/sktime/skpro/pull/648 - comments/comparison would be appreciated
* can you make the case for adding a ZINB and ZIPoisson explicitly (as a shorthand) instead of using the compositor?

**Comment by marrov** [comment]
> * there is a pull request for the compositor already, here: #648 - comments/comparison would be appreciated

Not sure why I missed this! I'll try to get around to adding some comments there

> * can you make the case for adding a ZINB and ZIPoisson explicitly (as a shorthand) instead of using the compositor?

In all honesty, because I needed them for work as ZINB and ZIP and classical distribution for intermittent demand so I thought that given that I had coded them up, I might as well contribute them. It allowed me an easier integration with `xgboostlss` and more potential for optimizing them individually. But I understand this may not be a strong argument.

**Comment by fkiraly** [comment]
> It allowed me an easier integration

I think that is a strong argument - specifically if it is a commonly seen distribution type. I also know it from intermittency modelling tbh.

File: skpro/tests/_config.py [hunk 1]
```diff
@@ -17,4 +17,17 @@
 ]
 
 
-EXCLUDED_TESTS = {}
+EXCLUDED_TESTS = {
+    # Zero-inflated distributions extend _ScipyAdapter but have different
```
Comment (fkiraly) [inline skpro/tests/_config.py:21]: should be in the tag `tests:skip_by_name` instead

**Review by fkiraly — CHANGES_REQUESTED** [review]
I have now merged #648, so you can work on the comparison/merge in this PR only, which should make that easier.

Current discrepancies that I see:

* #648 has broadcasting across `p` (good) while this PR does not
* this PR calls the parameter `pi` which is the same as `math.pi` (bad) , #648 has `p`
* the docstring details of this PR are better; the summary of #648 is better since it refers to composition
* shape validation on `p` does not need to be done if it is in `broadcast_params` as in #648 (good); this PR does it extra
* both PR implement the same methods, but different ways, so this needs to be checked
* `iloc` and `iat` are identical
* author credit should be moved to tag
